pkg/imageupdater/imageupdater.go

```diff
@@ -0,0 +1,187 @@
+package imageupdater
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/nais/knorten/pkg/chart"
+	"github.com/nais/knorten/pkg/database"
+	"github.com/nais/knorten/pkg/database/crypto"
+	"github.com/nais/knorten/pkg/database/gensql"
+	"github.com/nais/knorten/pkg/k8s"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	profileListHelmKey = "singleuser.profileList"
+)
+
+type ImageUpdater struct {
+	repo          *database.Repo
+	jupyterClient *chart.JupyterhubClient
+	log           *logrus.Entry
+}
+
+type garImage struct {
+	Name string `json:"package"`
+	Tag  string `json:"tags"`
+}
+
+type profile struct {
+	Name                string `json:"display_name"`
+	Description         string `json:"description"`
+	KubespawnerOverride struct {
+		Image string `json:"image"`
+	} `json:"kubespawner_override"`
+}
+
+func New(repo *database.Repo, k8sClient *k8s.Client, cryptClient *crypto.EncrypterDecrypter, jupyterChartVersion string, log *logrus.Entry) *ImageUpdater {
+	jupyterClient := chart.NewJupyterhubClient(repo, k8sClient, cryptClient, jupyterChartVersion, log.WithField("subsystem", "jupyterClient"))
+	return &ImageUpdater{
+		repo:          repo,
+		jupyterClient: &jupyterClient,
+		log:           log,
+	}
+}
+
+func (d *ImageUpdater) Run(frequency time.Duration) {
+	ctx := context.Background()
+
+	ticker := time.NewTicker(frequency)
+	defer ticker.Stop()
+	for {
+		d.run(ctx)
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+		}
+	}
+}
+
+func (d *ImageUpdater) run(ctx context.Context) {
+	profilesDB, err := d.repo.GlobalValueGet(ctx, gensql.ChartTypeJupyterhub, profileListHelmKey)
+	if err != nil {
+		d.log.WithError(err).Error("getting jupyterhub singleuser profiles from database")
+		return
+	}
+
+	profiles := []*profile{}
+	if err := json.Unmarshal([]byte(profilesDB.Value), &profiles); err != nil {
+		d.log.WithError(err).Error("unmarshalling profiles")
+		return
+	}
+
+	profilesStatus := []bool{}
+	for _, p := range profiles {
+		updated, err := updateIfNeeded(p)
+		if err != nil {
+			d.log.WithError(err).Error("checking image up to date with GAR")
+		}
+		profilesStatus = append(profilesStatus, updated)
+	}
+
+	if hasUpdates(profilesStatus) {
+		profilesB, err := json.Marshal(profiles)
```

Comment:
Hva betyr `profilesB`? Hva står `B`-en for?

Reply:
bytes, men endrer!